generator/build_integration.zig

```diff
@@ -63,18 +86,30 @@ pub const ShaderCompileStep = struct {
         const self = @fieldParentPtr(ShaderCompileStep, "step", step);
         const cwd = std.fs.cwd();
 
-        const cmd = try self.builder.allocator.alloc([]const u8, self.glslc_cmd.len + 3);
+        const cmd = try self.builder.allocator.alloc([]const u8, self.glslc_cmd.len + 5);
         for (self.glslc_cmd) |part, i| {
             cmd[i] = part;
         }
-        cmd[cmd.len - 2] = "-o";
 
         for (self.shaders.items) |shader| {
+            var argc: usize = cmd.len - 2;
+
+            if (shader.entry_point) |entry_point| {
+                cmd[argc - 3] = try std.fmt.allocPrint(self.builder.allocator, "-fentry-point={s}", .{entry_point});
+                argc += 1;
+            }
+
+            if (shader.stage) |stage| {
+                cmd[argc - 3] = try std.fmt.allocPrint(self.builder.allocator, "-fshader-stage={s}", .{@tagName(stage)});
+                argc += 1;
+            }
+
             const dir = path.dirname(shader.full_out_path).?;
             try cwd.makePath(dir);
-            cmd[cmd.len - 3] = shader.source_path;
-            cmd[cmd.len - 1] = shader.full_out_path;
-            try self.builder.spawnChild(cmd);
+            cmd[argc - 3] = shader.source_path;
+            cmd[argc - 2] = "-o";
+            cmd[argc - 1] = shader.full_out_path;
+            try self.builder.spawnChild(cmd[0..argc]);
```

Comment:
This part feels a bit ad-hoc with all the offsets now. I suggest to replace it with an array list:
```zig
var cmd = ArrayList([]const u8).init(self.builder.allocator);
try cmd.appendSlice(self.glslc_cmd);
const base_cmd_len = cmd.items.len;

for (self.shaders.items) |shader| {
  cmd.items.len = base_cmd_len;
  if (shader.entry_point) |entry_point| {
    try cmd.append(...);
  }
  ...
  try cmd.appendSlice(.{shader.source_path, "-o", shader.full_out_path});
  try self.builder.spawnChild(cmd.items);
}
```

Reply:
You're right, it started to feel very messy. I'll clean that up as you suggest.